### PR TITLE
optimize sampled_addmm performance on CPU (SparseCSR)

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91501
* #91500
* #91499
* __->__ #90978

### Target and Background
This PR is improving the performance of `sampled_addmm` on CPU device. This is part of effort for improving PyG performance on CPU for GNN training/inference.

The current implementation is a reference design which converts `SparseCSR` tensor back to dense tensor and then do the addmm and convert back to `SparseCSR` again: this is going to be very slow and won't be able to run most of the datasets under https://github.com/snap-stanford/ogb (convert to dense would trigger `OOM`).

### Benchmarks

Right now we don't have any hands-on benchmark or workload to test this since this operator is not used in PyG yet. I fetched the dataset from `ogb-products` where:

* number of nodes: 2.4 * 10^6
* number of edges: 1.26 * 10^8
* number of features: 128

So if we store the **adjacency matrix** is dense, it is going to be 2.4 * 2.4 * 4 * 10^12 bytes, this will be OOB on current code. I abstract the first 1k rows to compare, **1100x** speedup:

CPU: Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, dual socket, 20 cores per socket.
```
### before: run 1000 rows from the whole dataset
sampled_addmm: running dataset ogb-products first 1000 rows: each iter takes 1212.000 ms!

### after: run 1000 rows from the whole dataset
sampled_addmm: running dataset ogb-products first 1000 rows: each iter takes 1.102 ms!

### after: run the whole dataset
sampled_addmm: running dataset ogb-products (the whole dataset) 2449029 rows: each iter takes 873.306 ms!
```

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by mingfeima** [comment]
TODO list for 2nd stage of PyG performance optimization on CPU:

- [x] optimization of `sampled_addmm` on SparseCSR
- [x] enabling of `sampled_addm` on SparseCOO: canceled (`coalesce()` takes too much time, the above dataset would take `7.2s` to finish this)
- [ ] unify `ReduceTypes`: GNN would rely on a few operators who have similar `ReduceTypes`, such as `ScatterReduce`, `SegmentReduce`, `SampledReduce`, `SpmmReduce`
- [ ] optimization of `segment_reduce` with `lengths` and `offsets`
- [ ] migrate `sampled_reduce`
- [ ] enable multi aggregation

**Comment by mingfeima** [comment]
@rusty1s could you please review this one? `sampled_addmm` is mapped to SDDMM on cuda device but we don't have equivalent in MKL.

File: aten/src/ATen/native/cpu/SampledAddmmKernel.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  // usually, collapse B and M is a better option,
+  // but for most commonly used case (mat1 and mat2 is 2d tensor), B = 1,
+  // so choose to parallel on M so that we don't have to handle b offset frequently
+  using Vec = vec::Vectorized<scalar_t>;
+  utils::parallel_sparse_csr(crow_data, M, nnz, [&](int64_t begin, int64_t end) {
+    for (const auto b : c10::irange(B)) {
+      index_t* crow_ptr = crow_data + b * (M + 1);
+      index_t* col_ptr = col_data + b * nnz;
+      scalar_t* value_ptr = value_data + b * nnz;
```
Comment (pearu) [inline aten/src/ATen/native/cpu/SampledAddmmKernel.cpp:52]: https://github.com/pytorch/pytorch/pull/91243 added support for CSR indices contiguity per batch and non-contiguous values. Could we use an accessor here so that non-contiguous crow/col_indices and values in `result` could be supported?

Comment (nikitaved) [reply aten/src/ATen/native/cpu/SampledAddmmKernel.cpp:52]: In order to accomplish that one could use a TI interface to ease batch advancement. Looks like `cpu_kernel_vec` could be used here.

File: aten/src/ATen/native/sparse/SparseBlas.cpp [hunk 2]
```diff
@@ -132,8 +133,20 @@ Tensor& sparse_sampled_addmm_out_sparse_csr_cpu(
...
+  if (mat1.numel() == 0 || mat2.numel() == 0 || result._nnz() == 0) {
+    result.mul_(beta);
+    return result;
+  }
+
+  // transpose mat2 to [b, n, k] from performance perspective.
+  // for gnn classic usage, mat2 is already stored in [b, n, k] physically,
+  // so no extra memcpy is needed.
+  auto mat2_t = mat2.transpose(-1, -2).contiguous();
+  sampled_addmm_sparse_csr_stub(kCPU, mat1.contiguous(), mat2_t, beta, alpha, result);
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseBlas.cpp:148]: We should probably error out on hybrid inputs.

File: aten/src/ATen/native/cpu/SampledAddmmKernel.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/SampledAddmmKernel.cpp]: OK I see, with #91243 merged, we can't expect `values`, `crow` and `col` to be always contiguous. I have modified this PR to use `accessor` instead raw pointers.

Using TI for this one is not a good idea: if i use TI, it will be shape of {B, M} and squash {K}. The parallelization will be evenly done on {B, M} which is not the best choice from performance perspective. The reason is that on real GNN datasets, the edges each node have may vary quite a lot, consider the following example: say we have 10 nodes, and node0 has 10 edges, while node1~9 has only 1 edge, and we have 2 threads. TI will end up with `thread0: node0~4` and `thread1: 5~9`. The partition in this PR will have `thread0: node0` and `thread1:1~9`.

File: aten/src/ATen/native/sparse/SparseBlas.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:148]: Could you please specify: what do you mean by "hybrid inputs" here?

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:148]: Hybrid meaning with values with dim > 1.

File: aten/src/ATen/native/cpu/SampledAddmmKernel.cpp (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/cpu/SampledAddmmKernel.cpp]: Sorry, I am not sure I understand. It is possible to restride inputs whichever way it is beneficial for balanced workload (setting strides to zero and the corresponding dim sizes to suitable values). Given that `reshape` is used to squash batch dims, it is not that much different from the previous solution with calls to `contiguous` prior to kernel runs.

File: aten/src/ATen/native/sparse/SparseBlas.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:148]: The current check on `sampled_addmm` ( i did not change that) is to make sure `self` and `mat1` have the same batch size:

```C
TORCH_CHECK(
    !(self.dim() > 2 && self.sizes().slice(0, self.dim() - 2) != mat1.sizes().slice(0, mat1.dim() - 2)),
    "sampled_addmm: Expected self and mat1 to have the same batch size, but got ",
    self.sizes().slice(0, self.dim() - 2),
    " and ",
    mat1.sizes().slice(0, mat1.dim() - 2));
```

Will that be enough?

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:148]: Nope, hybrid tensors have shape `(batch_dims) + (sparse_dims) + (dense_dims)`. And the kernels you implemented assume no `dense_dims`.

Comment (pearu) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:148]: Adding
```c++
TORCH_CHECK(result.dense_dim() == 0, "sampled_addmm: Expected non-hybrid result tensor");
```
or similar to `sparse_sampled_addmm_check_inputs` ought to be sufficient to resolve this.

Comment (mingfeima) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:148]: Got it, thanks!

File: aten/src/ATen/native/sparse/SparseBlas.cpp [hunk 3]
```diff
@@ -172,6 +185,8 @@ void sparse_sampled_addmm_check_inputs(
       result.layout() == kSparseCsr,
       "sampled_addmm: Expected result to have sparse csr layout, but got ",
       result.layout());
+  TORCH_CHECK(result.dense_dim() == 0,
+      "sampled_addmm: Expected non-hybrid result tensor");
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:189]: I read "Expected non-hybrid result tensor" as "support for non-hybrid results not implemented yet", so it does not matter much if the exception is raised here or later after checking other inputs, it should fail anyway with hybrid `result`.

It could only matter if there exist tests that check the exception messages of invalid inputs but it is very likely that these tests would not use hybrid `result`.

That said, I would not mind moving the check further below where other checks on `result` are implemented.

[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:189]: Just a bit confusing to the user to read about `result` which is not supplied. And we could save on memory allocations if the inputs are already incorrect.

[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:189]: So, we need this check for both `self` and `result`.

[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/sparse/SparseBlas.cpp:191]: Oh thanks for the reminder! Updated!

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @mingfeima!

**Review by nikitaved — COMMENTED** [review]
Looks great! Thank you, @mingfeima !

**Review by cpuhrsch — APPROVED** [review]

**Comment by seemethere** [comment]
Would suggest making the following change when attempting to re-land:

```diff
diff --git a/build_variables.bzl b/build_variables.bzl
index 225fb9fbc7e..0359c5123c7 100644
--- a/build_variables.bzl
+++ b/build_variables.bzl
@@ -1430,6 +1430,7 @@ aten_native_source_non_codegen_list = [
     "aten/src/ATen/native/nested/NestedTensorUnaryOps.cpp",
     "aten/src/ATen/native/nested/NestedTensorUtils.cpp",
     "aten/src/ATen/native/sparse/ParamUtils.cpp",
+    "aten/src/ATen/native/cpu/SampledAddmmKernel.cpp",
     "aten/src/ATen/native/sparse/SoftMax.cpp",
     "aten/src/ATen/native/sparse/SparseBlas.cpp",
     "aten/src/ATen/native/sparse/SparseBlasImpl.cpp",
```

**Comment by huydhn** [comment]
Thank @seemethere for reverting this one, I'm about to do the same because this also breaks periodic `buck-build-test` https://hud.pytorch.org/commit/pytorch/pytorch/364f526b9cdf9818a7647b5e637efdee825d61a1

**Comment by ezyang** [comment]
Why is this not caught in oss build

**Comment by huydhn** [comment]
> Why is this not caught in oss build

It was caught by periodic buck build https://github.com/pytorch/pytorch/actions/runs/3889778963/jobs/6638325506.  Now with `ciflow/periodic` in the PR, the failure has shown up correctly

**Comment by mingfeima** [comment]
`build_variables.bzl` updated!
